### PR TITLE
Add the core/v3 API, minor improvements to core/v2

### LINKED ISSUE #3795: Create internal core/v3 API for entities
Create a new core/v3 API that includes EntityConfig and EntityState. As part of this objective, create a function that allows a user to create a core/v2.Entity by combining a core/v3.EntityConfig and a core/v3.EntityState. 

Background: design doc https://docs.google.com/document/d/1p43VdkuSU4_lndiVZ_K-5cQryEpNCAOBmVeaqmAaubk/edit

Generate the protobufs into Go structs, and create a new core/v3 API package. Make sure to add all the necessary boilerplate and machinery necessary to dynamically resolve core/v3 datatypes.

Create a shim function for converting `core/v3` entity data into a `core/v2.Entity`:

```
func MakeCoreV2Entity(cfg *EntityConfig, state *EntityState) (*corev2.Entity, error) { ... }
```

Labels and annotations from `EntityConfig` and `EntityState` should be merged, with `EntityConfig` labels and annotations taking precedence. If namespace and name disagree between `EntityState` and `EntityConfig`, then an error should be returned.

A rough sketch of the data structures (may not compile):
```
syntax = "proto3";

import "github.com/gogo/protobuf@v1.3.1/gogoproto/gogo.proto";
import "github.com/sensu/sensu-go/api/core/v2/meta.proto";
import "github.com/sensu/sensu-go/api/core/v2/entity.proto";

package sensu.core.v3;

// TODO add docs
message EntityConfig {
    v2.ObjectMeta metadata = 1 [
	(gogoproto.jsontag) = "metadata,omitempty",
	(gogoproto.embed) = true,
	(gogoproto.nullable) = false
    ];

    string entity_class = 2 [(gogoproto.jsontag) = "entity_class"];

    string user = 3;

    repeated string subscriptions = 4 [(gogoproto.jsontag) = "subscriptions"];

    bool deregister = 5 [(gogoproto.jsontag) = "deregister"];

    v2.Deregistration deregistration = 6 [(gogoproto.nullable) = false];

    repeated string keepalive_handlers = 7;

    repeated string redact = 8;
}

// TODO add docs
message EntityState {
    v2.ObjectMeta metadata = 1 [
	(gogoproto.jsontag) = "metadata,omitempty",
	(gogoproto.embed) = true,
	(gogoproto.nullable) = false
    ];

    v2.System system = 2 [(gogoproto.nullable) = false];

    int64 last_seen = 3 [(gogoproto.jsontag) = "last_seen"];

    string sensu_agent_version = 4
	[(gogoproto.jsontag) = "sensu_agent_version"];
}
```

### PR DESCRIPTION
**Pull request by echlebek**
## What is this change?
Note: I've CC'd all other engineers on this change in order to make sure everyone has a chance to see it. Not everyone needs to do a full code review (although everyone is certainly welcome to!).

```
commit cd93ba2bc3bf4e05c19e22c7e137c8ddc79753f0 (HEAD -> core/v3)
Author: Eric Chlebek <eric@sensu.io>
Date:   Tue May 26 15:26:44 2020 -0700

    Add core/v3 API
    
    This commit adds the beginnings of the core/v3 API. It includes
    the EntityConfig and EntityState resources.
    
    The core/v3 API expands on the ideas introduced in the core/v2 API,
    and refines the Resource concept. A new core/v3.Resource interface
    has been defined, and it's expected that all resources in the
    core/v3 package will conform to this interface.
    
    We've had a lot of experience developing the core/v2 API over the
    last couple of years, and as a result it's been possible to
    generalize that experience into code generation. Developers can
    now create core/v3.Resource implementations by doing nothing more
    than authoring a .proto file.
    
    The automatically generated code has automatically generated tests
    to go along with it, which should ensure that changes and updates
    to the code generation over time will not cause breakage.
    
    UnmarshalJSON has been automatically generated for all types, so
    that the common pitfall of nil labels and annotations can be
    avoided. All types that have ObjectMeta are therefore guaranteed
    to have non-nil labels and annotations after being unmarshalled
    from JSON.
    
    Most of the automatically generated methods that implement Resource
    can be overridden by writing a custom unexported method for the
    type in question. See the code documentation for more details.
    
    Validation works more cooperatively, with the automatically
    generated method only checking the metadata, and then optionally
    invoking the unexported validate() method, if it is present.
    
    Any core/v3.Resource can be converted to a core/v2.Resource with
    the V3ResourceToV2 function. This should allow for gradual code
    repair throughout the sensu-go codebase over time.
    
    The EntityConfig and EntityState types are a split of the original
    core/v2.Entity type. These types have been introduced because we
    want to start treating these data different from a storage
    perspective.
    
    Functions have been included to convert between core/v2.Entity, and
    core/v3.{EntityConfig, EntityState}. This will allow developers to
    gradually use EntityConfig and EntityState in the codebase without
    worrying too much about compatibility.
    
    Signed-off-by: Eric Chlebek <eric@sensu.io>

commit 4fb151958d0962072fae21d0f5d1ec5f53fa724d
Author: Eric Chlebek <eric@sensu.io>
Date:   Tue May 26 15:16:04 2020 -0700

    Refactor core/v2 protos
    
    This commit refactors the import paths of the core/v2 protos so
    that they're easier to use from outside of the package.
    
    Signed-off-by: Eric Chlebek <eric@sensu.io>
```

## Why is this change necessary?

Closes #3795 

## Does your change need a Changelog entry?

Yes

## Were there any complications while making this change?

I had to refactor the import paths of the core/v2 protos before they could be used outside their own package. This is because the protos were using relative import paths.

## Have you reviewed and updated the documentation for this change? Is new documentation required?

This API is not ready for publication, and as such, no docs changes are desired.

## How did you verify this change?

Unit tests only so far. I expect that bugs will be fixed as the API is integrated into various parts of the system.

## Is this change a patch?

No

### CODE REVIEW

File: api/core/v3/entityconfig.proto [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+option go_package = "v3";
+option (gogoproto.populate_all) = true;
+option (gogoproto.equal_all) = true;
+option (gogoproto.marshaler_all) = true;
+option (gogoproto.unmarshaler_all) = true;
+option (gogoproto.sizer_all) = true;
+option (gogoproto.testgen_all) = true;
+
+// EntityConfig represents entity configuration.
+message EntityConfig {
```
Comment (nikkictl) [inline api/core/v3/entityconfig.proto:18]: Quite a minor comment, but isn't it more stylistically correct (and consistent) to name this file `entity_config.proto`?

File: api/core/v3/entitystate.proto [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+option (gogoproto.equal_all) = true;
+option (gogoproto.marshaler_all) = true;
+option (gogoproto.unmarshaler_all) = true;
+option (gogoproto.sizer_all) = true;
+option (gogoproto.testgen_all) = true;
+
+// EntityState represents entity state. Unlike configuration, state is
+// typically only maintained for agent entities, although it can be maintained
+// for proxy entities in certain circumstances.
+message EntityState {
```
Comment (nikkictl) [inline api/core/v3/entitystate.proto:20]: Same here for `entity_state.proto`?

File: api/core/v3/resource_generated_test.go [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+
+func TestEntityConfigSetMetadata(t *testing.T) {
+	value := new(EntityConfig)
+	var iface interface{} = value
+	metaer, ok := iface.(getMetadataer)
+	if !ok {
+		return
+	}
+	meta := &corev2.ObjectMeta{
+		Name:        "snoopdogg",
```
Comment (nikkictl) [inline api/core/v3/resource_generated_test.go:22]: giphy

File: api/core/v3/typemap.go [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+// ResolveResource returns a zero-valued resource, given a name.
+// If the named type does not exist, or if the type is not a Resource,
+// then an error will be returned.
+func ResolveResource(name string) (Resource, error) {
+	t, ok := typeMap[name]
+	if !ok {
+		return nil, fmt.Errorf("type could not be found: %q", name)
+	}
+	if _, ok := t.(Resource); !ok {
+		return nil, fmt.Errorf("%q is not a Resource", name)
```
Comment (nikkictl) [inline api/core/v3/typemap.go:48]:
```suggestion
		return nil, fmt.Errorf("%q is not a v3 Resource", name)
```

File: scripts/generate_type/generate_type.go [hunk 5]
```diff
@@ -21,8 +21,14 @@ var (
 	typeRe   = regexp.MustCompile(`^type ([A-Z].+) struct \{`)
 )
 
-type typeNames struct {
+type tmplData struct {
 	TypeNames []string
+	Comment   string
+}
+
+// kebabCase is like snakeCase, but underscores are converted to dashes
```
Comment (nikkictl) [inline scripts/generate_type/generate_type.go:29]: TIL

File: api/core/v3/typemap.go (hunk 4 above)
Comment (echlebek) [reply api/core/v3/typemap.go]: :+1:

File: api/core/v3/entitystate.proto (hunk 2 above)
Comment (echlebek) [reply api/core/v3/entitystate.proto]: :+1:

**Pull request merged**